Database connection pool metrics semantic conventions (#2273)
open-telemetry/opentelemetry-specification: specification/metrics/semantic_conventions/database-metrics.md

# General

**Status**: [Experimental](../../document-status.md)

The conventions described in this section are specific to SQL and NoSQL clients.

**Disclaimer:** These are initial database client metric instruments and attributes but more may be added in the future.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Metric Instruments](#metric-instruments)
  * [Connection pools](#connection-pools)

<!-- tocstop -->

## Metric Instruments

The following metric instruments MUST be used to describe database client operations. They MUST be of the specified type
and units.

### Connection pools

> Below is a table of database connection pool server metric instruments that MUST be used by connection pool

Reviewer: Are these metric events for a database server to measure and report? If so, [this 
line](https://github.com/open-telemetry/opentelemetry-specification/pull/2273/files#diff-dd95ea8b8c08c0de9f797026a77365fc91472ab8f87f6e3f87ac60107d43ff2eR7) needs to be updated to state this document includes both server and client conventions.

Author: Oh, my bad - these are purely client side metrics.

instrumentations:

| Name                        | Instrument                 | Unit         | Unit ([UCUM](README.md#instrument-units)) | Description |
|-----------------------------|----------------------------|--------------|-------------------------------------------|-------------|

> | `db.connection_pool.total`  | Asynchronous UpDownCounter | connections  | `{connections}`                           | The total number of open connections.

Reviewer: Can we rename `connection_pool` to just `connections`?

| `db.connection_pool.active` | Asynchronous UpDownCounter | connections  | `{connections}`                           | The number of open connections that are currently in use.
| `db.connection_pool.idle`   | Asynchronous UpDownCounter | connections  | `{connections}`                           | The number of open connections that are currently idle.

> Connection pool instrumentations SHOULD use the following metric instruments, if applicable:

Reviewer: The way this normative statement is made it could be interpreted as "if the data is available, use these instruments", but it could also be interpreted as "if the developer wants to, use these instruments". This latter interpretation could lead to compatibility issues.  Can we rephrase this so only the first interpretation is communicated?

i.e.

```suggestion
Instrumentation libraries for database server connection pools that collect data for the following data MUST use the following metric instruments. Otherwise, if the instrumentation library does not collect this data, these instruments MUST NOT be used.
```

Author: Yep, the first interpretation is what I aimed for

| Name                                 | Instrument                 | Unit         | Unit ([UCUM](README.md#instrument-units)) | Description |
|--------------------------------------|----------------------------|--------------|-------------------------------------------|-------------|
| `db.connection_pool.idle.max`        | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of idle open connections allowed.
| `db.connection_pool.idle.min`        | Asynchronous UpDownCounter | connections  | `{connections}`                           | The minimum number of idle open connections allowed.
| `db.connection_pool.max`             | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of open connections allowed.

> | `db.connection_pool.pending_threads` | Asynchronous UpDownCounter | threads      | `{threads}`                               | The number of threads that are currently waiting for an open connection.

Reviewer: Threads is not a universal term across OpenTelemetry languages. Naming this as just `db.connection_pool.pending` would have a better chance at compatibility. Additionally, using units of `connections` would describe more universally the quantity of pending things.

Reviewer: > How about `pending_requests`/`requests`? This instruments is supposed to represents the number of threads waiting for an idle connection - or a number of requests for an idle connection from the pool.

SGTM

Author: I used `waiting_requests` in the end - to match the `db.connections.time[operation = wait]` instrument & attribute name

| `db.connection_pool.timeouts`        | Counter                    | timeouts     | `{timeouts}`                              | The number of connection timeouts that have happened since the application start.
| `db.connection_pool.create_time`     | Histogram                  | milliseconds | `ms`                                      | The time it took to create a new connection.
| `db.connection_pool.wait_time`       | Histogram                  | milliseconds | `ms`                                      | The time it took to get an open connection from the pool.

> | `db.connection_pool.use_time`        | Histogram                  | milliseconds | `ms`                                      | The time between borrowing a connection and returning it to the pool.

Reviewer: Our [instrument naming guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/e4d0c4c1e2474f89e3d9f58baaa0afcd7d9ac075/specification/metrics/semantic_conventions/README.md#instrument-naming) recommend collapsing all three of these into a single `db.connection_pool.time` instrument.

> Below is a table of the attributes that are to be included on connection pool metric events:

Reviewer: Can we specify this normatively?

```suggestion
Below is a table of the attributes that MUST be included on connection pool metric events:
```

# General

**Status**: [Experimental](../../document-status.md)

The conventions described in this section are specific to SQL and NoSQL clients.

**Disclaimer:** These are initial database client metric instruments and attributes but more may be added in the future.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Metric Instruments](#metric-instruments)
  * [Connection pools](#connection-pools)

<!-- tocstop -->

## Metric Instruments

The following metric instruments MUST be used to describe database client operations. They MUST be of the specified type
and units.

### Connection pools

Below is a table of database client connection pool metric instruments that MUST be used by connection pool
instrumentations:

| Name                          | Instrument                 | Unit        | Unit ([UCUM](README.md#instrument-units)) | Description                                                                               |
|-------------------------------|----------------------------|-------------|-------------------------------------------|-------------------------------------------------------------------------------------------|

> | `db.client.connections.usage` | Asynchronous UpDownCounter | connections | `{connections}`                           | The number of connections that are currently in state described by the `state` attribute. |

Reviewer 2: (based on recent change in #2458)
```suggestion
| `db.client.connections.usage` | UpDownCounter | connections | `{connections}`                           | The number of connections that are currently in state described by the `state` attribute. |
```

| `db.connections.active` | Asynchronous UpDownCounter | connections  | `{connections}`                           | The number of open connections that are currently in use.

> | `db.connections.idle`   | Asynchronous UpDownCounter | connections  | `{connections}`                           | The number of open connections that are currently idle.

Author: Nice, I like that. I renamed the attribute to `db.connections.usage`, cause that seems to match the general naming conventions better.

# General

**Status**: [Experimental](../../document-status.md)

The conventions described in this section are specific to SQL and NoSQL clients.

**Disclaimer:** These are initial database client metric instruments and attributes but more may be added in the future.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Metric Instruments](#metric-instruments)
  * [Connection pools](#connection-pools)

<!-- tocstop -->

## Metric Instruments

The following metric instruments MUST be used to describe database client operations. They MUST be of the specified type
and units.

### Connection pools

Below is a table of database client connection pool metric instruments that MUST be used by connection pool

> instrumentations:

Reviewer 3: We should probably disambiguate between the metrics for db client connection pool and db server connection pool monitoring. E.g. `db.client.connections.usage`, `db.server.connections.idle.max`.

Author: Oh, good point 👍

| Name                   | Instrument                 | Unit        | Unit ([UCUM](README.md#instrument-units)) | Description |
|------------------------|----------------------------|-------------|-------------------------------------------|-------------|
| `db.connections.usage` | Asynchronous UpDownCounter | connections | `{connections}`                           | The number of connections that are currently in state described by the `state` attribute.

All `db.connections.usage` measurements MUST include the following attribute:

| Name    | Type   | Description                                                                  | Examples | Required |
|---------|--------|------------------------------------------------------------------------------|----------|----------|
| `state` | string | The state of a connection in the pool. Valid values include: `idle`, `used`. | `idle`   | Yes      |

> Instrumentation libraries for database server connection pools that collect data for the following data MUST use the

Reviewer 4: Why are the basic R.E.D. metrics not captured? only the D is present as *.time

Author: I have found these metrics to be already available in most connection pool libraries; the libraries that I've worked with rarely provide anything more.

following metric instruments. Otherwise, if the instrumentation library does not collect this data, these instruments
MUST NOT be used.

| Name                              | Instrument                 | Unit         | Unit ([UCUM](README.md#instrument-units)) | Description |
|-----------------------------------|----------------------------|--------------|-------------------------------------------|-------------|
| `db.connections.idle.max`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of idle open connections allowed.
| `db.connections.idle.min`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The minimum number of idle open connections allowed.
| `db.connections.max`              | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of open connections allowed.

> | `db.connections.waiting_requests` | Asynchronous UpDownCounter | requests     | `{requests}`                              | The number of pending requests for an open connection.

Reviewer 4: "For an open connection" sounds like some connection id must be used as an attribute, which I assume is not the case. Please rephrase. Is that a cumulative number across all open connections?

Author: Yes, that's cumulative.

> | `db.connections.timeouts`         | Counter                    | timeouts     | `{timeouts}`                              | The number of connection timeouts that have happened since the application start.

Reviewer 4: Are these timeouts when establishing a connection or when executing data operations?

Author: These are meant to be timeouts when trying to obtain a connection from the pool -- I'll add it to the description, thanks.

> | `db.client.connections.waiting_requests` | Asynchronous UpDownCounter | requests     | `{requests}`                              | The number of pending requests for an open connection, cumulative for the entire pool.                                                       |

Author: I changed it to `pending_requests`

| `db.connections.time`             | Histogram                  | milliseconds | `ms`                                      | The time it took to apply an operation described by the `operation` attribute.

All `db.connections.time` measurements MUST include the following attribute:

| Name        | Type   | Description                                                                                   | Examples | Required |
|-------------|--------|-----------------------------------------------------------------------------------------------|----------|----------|

> | `operation` | string | The type of operation applied to a connection. Valid values include: `create`, `use`, `wait`. | `create` | Yes      |

Reviewer 4: The meaning of the listed operations is not clear to me, can they be explained?

Author: Done.

> The `db.connections.usage` metric events MUST include the following attribute:
> 
> | Name    | Type   | Description                                                                  | Examples | Required |
> |---------|--------|------------------------------------------------------------------------------|----------|----------|
> | `state` | string | The state of a connection in the pool. Valid values include: `idle`, `used`. | `idle`   | Yes      |
> 
> Instrumentation libraries for database server connection pools that collect data for the following data MUST use the
> following metric instruments. Otherwise, if the instrumentation library does not collect this data, these instruments
> MUST NOT be used.
> 
> | Name                              | Instrument                 | Unit         | Unit ([UCUM](README.md#instrument-units)) | Description |
> |-----------------------------------|----------------------------|--------------|-------------------------------------------|-------------|
> | `db.connections.idle.max`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of idle open connections allowed.
> | `db.connections.idle.min`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The minimum number of idle open connections allowed.
> | `db.connections.max`              | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of open connections allowed.
> | `db.connections.waiting_requests` | Asynchronous UpDownCounter | requests     | `{requests}`                              | The number of pending requests for an open connection.
> | `db.connections.timeouts`         | Counter                    | timeouts     | `{timeouts}`                              | The number of connection timeouts that have happened since the application start.
> | `db.connections.time`             | Histogram                  | milliseconds | `ms`                                      | The time it took to apply an operation described by the `operation` attribute.
> 
> The `db.connections.time` metric events MUST include the following attribute:
> 
> | Name        | Type   | Description                                                                                   | Examples | Required |
> |-------------|--------|-----------------------------------------------------------------------------------------------|----------|----------|
> | `operation` | string | The type of operation applied to a connection. Valid values include: `create`, `use`, `wait`. | `create` | Yes      |
> 
> Below is a table of the attributes that MUST be included on all connection pool metric events:

Reviewer 5: "event" could be misleading, "measurement" would be the proper term as per https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#measurement.

```suggestion
All `db.connections.usage` measurements MUST include the following attribute:

| Name    | Type   | Description                                                                  | Examples | Required |
|---------|--------|------------------------------------------------------------------------------|----------|----------|
| `state` | string | The state of a connection in the pool. Valid values include: `idle`, `used`. | `idle`   | Yes      |

Instrumentation libraries for database server connection pools that collect data for the following data MUST use the
following metric instruments. Otherwise, if the instrumentation library does not collect this data, these instruments
MUST NOT be used.

| Name                              | Instrument                 | Unit         | Unit ([UCUM](README.md#instrument-units)) | Description |
|-----------------------------------|----------------------------|--------------|-------------------------------------------|-------------|
| `db.connections.idle.max`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of idle open connections allowed.
| `db.connections.idle.min`         | Asynchronous UpDownCounter | connections  | `{connections}`                           | The minimum number of idle open connections allowed.
| `db.connections.max`              | Asynchronous UpDownCounter | connections  | `{connections}`                           | The maximum number of open connections allowed.
| `db.connections.waiting_requests` | Asynchronous UpDownCounter | requests     | `{requests}`                              | The number of pending requests for an open connection.
| `db.connections.timeouts`         | Counter                    | timeouts     | `{timeouts}`                              | The number of connection timeouts that have happened since the application start.
| `db.connections.time`             | Histogram                  | milliseconds | `ms`                                      | The time it took to apply an operation described by the `operation` attribute.

All `db.connections.time` measurements MUST include the following attribute:

| Name        | Type   | Description                                                                                   | Examples | Required |
|-------------|--------|-----------------------------------------------------------------------------------------------|----------|----------|
| `operation` | string | The type of operation applied to a connection. Valid values include: `create`, `use`, `wait`. | `create` | Yes      |

Below is a table of the attributes that MUST be included on all connection pool measurements:
```

> | `db.client.connections.time`             | Histogram                  | milliseconds | `ms`                                      | The time it took to apply an operation described by the `operation` attribute.                                                               |

Reviewer 2: ping @MrAlias @mateuszrzeszutek

> | `db.client.connections.time`             | Histogram                  | milliseconds | `ms`                                      | The time it took to apply an operation described by the `operation` attribute.                                                               |

Author: @MrAlias I reverted this change back to 3 separate instruments, PTAL.

| Name        | Type   | Description                                                                  | Examples       | Required |
|-------------|--------|------------------------------------------------------------------------------|----------------|----------|

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. | `myDataSource` | Yes      |

Reviewer 4: What if naming of the pools is not supported by the connection pool library?

Author: The name of the pool is rather important to distinguish between differrent connection pools; if the instrumented library does not provide it, then I believe that the user of the instrumentation should. But then again, all libraries I've worked with had the notion of a pool name.
